sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala

```diff
@@ -0,0 +1,317 @@
+      .setCalendarType("gregory")
+      .setInstant(microsToMillis(micros))
+      .build()
+    val localDateTime = LocalDateTime.of(
+      cal.get(Calendar.YEAR),
+      cal.get(Calendar.MONTH) + 1,
+      // The number of days will be added later to handle non-existing
+      // Julian dates in Proleptic Gregorian calendar.
+      // For example, 1000-02-29 exists in Julian calendar because 1000
+      // is a leap year but it is not a leap year in Gregorian calendar.
+      1,
+      cal.get(Calendar.HOUR_OF_DAY),
+      cal.get(Calendar.MINUTE),
+      cal.get(Calendar.SECOND),
+      (Math.floorMod(micros, MICROS_PER_SECOND) * NANOS_PER_MICROS).toInt)
+      .plusDays(cal.get(Calendar.DAY_OF_MONTH) - 1)
+    val zonedDateTime = localDateTime.atZone(zoneId)
+    // Zero DST offset means that local clocks have switched to the winter time already.
+    // So, clocks go back one hour. We should correct zoned date-time and change
+    // the zone offset to the later of the two valid offsets at a local time-line overlap.
+    val adjustedZdt = if (cal.get(Calendar.DST_OFFSET) == 0) {
+      zonedDateTime.withLaterOffsetAtOverlap()
+    } else {
+      zonedDateTime
+    }
+    instantToMicros(adjustedZdt.toInstant)
+  }
+
+  // The rebasing maps to convert microseconds from the hybrid calendar (Julian + Gregorian)
+  // to Proleptic Gregorian calendar.
+  // `julianGregDiffSwitchMicros` maps time zone IDs to ordered timestamps (ascending order)
+  // where at every timestamps the difference between 2 calendars was changed.
+  // `julianGregDiffsMicros` maps time zone IDs to ordered differences between 2 calendars.
+  // The diff at the index `i` is applicable for all timestamps in the interval:
+  // [julianGregDiffSwitchMicros(i), julianGregDiffSwitchMicros(i+1))
+  private val julianGregRebaseMap = loadRebaseRecords("julian-gregorian-rebase-micros.json")
+
+  /**
+   * This is an opposite to `rebaseGregorianToJulianMicros` function which rebases the given
+   * microseconds since the epoch 1970-01-01T00:00:00.000000Z via local timestamps from the
+   * hybrid calendar (Julian + Gregorian) to Proleptic Gregorian calendar.
+   * For example, the input `micros` -12243196799876544 is mapped to 1582-01-01 00:00:00.123456 in
+   * Julian calendar. The same local timestamp in Proleptic Gregorian calendar is mapped to
+   * the different number of micros -12244061221876544 since the epoch. Semantically, the function
+   * performs such conversion either via direct conversion to local timestamps,
+   * or via pre-calculated rebasing tables.
+   *
+   * @param micros The number of microseconds since the epoch.
+   * @return The rebased number of microseconds since the epoch which is mapped to the same
+   *         local timestamp in Proleptic Gregorian calendar as `micros` in the hybrid calendar
+   *         at the system time zone.
+   */
+  def rebaseJulianToGregorianMicros(micros: Long): Long = {
+    val timeZone = TimeZone.getDefault
```

Comment:
We need to add a bit more document that we assume julian seconds was generated with the JVM system default timezone.

Reply:
ditto for `rebaseGregorianToJulianMicros`

Reply:
Added notes